optaplanner-examples/src/main/java/org/optaplanner/examples/nurserostering/optional/score/NurseRosteringConstraintProvider.java

```diff
@@ -0,0 +1,480 @@
+                .join(Employee.class,
+                        Joiners.equal(MinMaxContractLine::getContract, Employee::getContract))
+                .join(ShiftAssignment.class,
+                        Joiners.equal((contractLine, employee) -> employee, ShiftAssignment::getEmployee))
+                .groupBy((line, employee, shift) -> ImmutablePair.of(employee, line), ConstraintCollectors.countTri())
+                .filter((employeeContractPair, shiftCount) -> employeeContractPair.getRight().isViolated(shiftCount))
+                .penalize("org.optaplanner.examples.nurserostering.solver", "Minimum and maximum number of assignments",
+                        HardSoftScore.ONE_SOFT,
+                        (employeeContractPair, shiftCount) -> employeeContractPair.getRight().getViolationAmount(shiftCount));
+    }
+
+    Constraint minimumNumberOfAssignmentsNoAssignments(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(MinMaxContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine.getContractLineType() == ContractLineType.TOTAL_ASSIGNMENTS &&
+                        minMaxContractLine.isEnabled())
+                .join(Employee.class,
+                        Joiners.equal(MinMaxContractLine::getContract, Employee::getContract))
+                .ifNotExists(ShiftAssignment.class,
+                        Joiners.equal((contractLine, employee) -> employee, ShiftAssignment::getEmployee))
+                .filter((contract, employee) -> contract.isViolated(0))
+                .penalize("org.optaplanner.examples.nurserostering.solver",
+                        "Minimum and maximum number of assignments (no assignments)", HardSoftScore.ONE_SOFT,
+                        (contract, employee) -> contract.getViolationAmount(0));
+    }
+
+    // Min/Max consecutive working days
+    // These Min/Max constraints are implemented as two constraints for consistency with DRL
+    Constraint minimumConsecutiveWorkingDays(ConstraintFactory constraintFactory) {
+        return getConsecutiveShifts(constraintFactory.from(MinMaxContractLine.class)
+                .filter(minMaxContractLine -> minMaxContractLine
+                        .getContractLineType() == ContractLineType.CONSECUTIVE_WORKING_DAYS &&
+                        minMaxContractLine.isMinimumEnabled()))
+                                .filter((employeeContractPair,
+                                        shiftList) -> shiftList.getLength() < employeeContractPair.getRight().getMinimumValue())
+                                .penalize("org.optaplanner.examples.nurserostering.solver", "minimumConsecutiveWorkingDays",
+                                        HardSoftScore.ONE_SOFT,
+                                        (employeeContractPair, shiftList) -> employeeContractPair.getRight()
+                                                .getViolationAmount(shiftList.getLength()));
```

Comment:
This is hard to read. The fact that the stream is passed to a function is fine, but I think that the stream should first be declared, and then the variable passed. Easier to understand at first sight.